Cache mapped functions

majosm opened the pull request:
Caches the results of mapping `FunctionDefinition`s. Caching is done globally to avoid re-traversing functions in different call stack frames.

Depends on ~~#503~~ (merged) and ~~#530~~ (merged).

majosm commented:
@inducer Should be ready for a look. I think the mypy errors are unrelated, they're occurring on main too.

inducer commented:
Uh-oh. The recent typing apocalypse has made a bit of a mess of this. I've started resolving conflicts, but I got bogged down in `pytato.transform`. Could you take a look? (Or, if you've lost patience with me, I can also try and finish.)

majosm commented:
@inducer Should be ready for review again (mypy error looks unrelated?). In e57fe22309fffa47fd86c779830e39b21eaa19bf I added a few questions/comments about things I noticed while resolving conflicts. I'll remove them after you've had a chance to glance at them.

inducer commented:
https://github.com/inducer/pytato/pull/571 should address the mypy issue.

inducer commented on pytato/transform/__init__.py:
```diff
@@ -218,11 +217,25 @@ def rec(self, expr: ArrayOrNames, *args: P.args, **kwargs: P.kwargs) -> ResultT:
+        """Call the mapper method of *expr* and return the result."""
+        method: Callable[..., Any] | None
+
+        try:
+            method = self.map_function_definition  # type: ignore[attr-defined]
+        except AttributeError:
+            return cast("FunctionResultT", self.map_foreign(expr, *args, **kwargs))
```
Can this be an error instead?

majosm replied:
Done.

inducer commented on pytato/transform/__init__.py:
```diff
@@ -194,7 +193,7 @@ def handle_unsupported_array(self, expr: MappedT,
                 f"{type(self).__name__} cannot handle expressions of type {type(expr)}")
 
-    def map_foreign(self, expr: Any, *args: P.args, **kwargs: P.kwargs) -> ResultT:
+    def map_foreign(self, expr: Any, *args: P.args, **kwargs: P.kwargs) -> Any:
```
I think we can get rid of `map_foreign`.

majosm replied:
`Reprifier` appears to override `map_foreign` to handle stringifying non-array fields. Seems like it needs to stay?

inducer replied:
Let's move it to `Reprifier` then. I don't think it needs to be a feature of the generic machinery.

majosm replied:
Done.

inducer commented on pytato/transform/__init__.py:
```diff
@@ -233,23 +246,42 @@ def __call__(self,
+        else:
+            function_cache = {}
+
+        self._function_cache: dict[Hashable, FunctionResultT] = function_cache
```
Can this be an `if` expression?

inducer commented on pytato/transform/__init__.py:
```diff
@@ -259,63 +291,68 @@ def rec(self, expr: ArrayOrNames, *args: P.args, **kwargs: P.kwargs) -> ResultT:
 class TransformMapperWithExtraArgs(
-            CachedMapper[ArrayOrNames, P],
-            Mapper[ArrayOrNames, P]
+            # Why multiple inheriting?
+            CachedMapper[ArrayOrNames, FunctionDefinition, P],
+            Mapper[ArrayOrNames, FunctionDefinition, P]
```
```suggestion
            CachedMapper[ArrayOrNames, FunctionDefinition, P],
```
I don't think it should be.

majosm replied:
Seems to work fine without it.

inducer commented on pytato/transform/__init__.py:
```diff
@@ -259,63 +291,68 @@ def rec(self, expr: ArrayOrNames, *args: P.args, **kwargs: P.kwargs) -> ResultT:
-
-    .. automethod:: clone_for_callee
     """
+    # FIXME: This inflates recursion depth
```
```
if TYPE_CHECKING:
    ... (same)
else:
    @property
    def rec_ary(self):
        return self.rec
```

inducer commented on pytato/transform/__init__.py:
```diff
@@ -259,63 +291,68 @@ def rec(self, expr: ArrayOrNames, *args: P.args, **kwargs: P.kwargs) -> ResultT:
-
-    .. automethod:: clone_for_callee
     """
+    # FIXME: This inflates recursion depth
```
```
def is_array(expr: ArrayOrNames) -> Array:
    assert isinstance(expr, Array)
    return expr
```
(This is probably the saner way.)

majosm replied:
Something like https://github.com/inducer/pytato/pull/531/commits/ca4f3366c669085b155f11a04f8cd30b07223b45?

(The other way appears to work too: https://github.com/majosm/pytato/commit/34af8ef15414b192c8259ea8b30e18456f4415fd)

inducer commented on pytato/transform/__init__.py:
```diff
@@ -744,6 +780,9 @@ class CombineMapper(Mapper[ResultT, []]):
     def __init__(self) -> None:
         super().__init__()
         self.cache: dict[ArrayOrNames, ResultT] = {}
+        # Don't need to pass function cache as argument here, because unlike
+        # CachedMapper we're not creating a new mapper for each call
+        self.function_cache: dict[FunctionDefinition, FunctionResultT] = {}
```
I think the behavior should be consistent across mapper types.

majosm replied:
Added `_function_cache` to the `__init__` args like the others. I didn't go the whole way and add `clone_for_callee`, etc., yet because I'm worried that's going to make a mess of my downstream changes.

inducer commented on pytato/transform/__init__.py:
```diff
@@ -1255,24 +1300,46 @@ class CachedWalkMapper(WalkMapper[P]):
+        else:
+            visited_functions = set()
+
+        self._visited_functions: set[Any] = visited_functions
```
`if` expr?

inducer commented on pytato/target/loopy/codegen.py:
```diff
@@ -384,7 +384,7 @@ def update_t_unit(self, t_unit: lp.TranslationUnit) -> None:
 # {{{ codegen mapper
 
-class CodeGenMapper(Mapper[ImplementedResult, [CodeGenState]]):
+class CodeGenMapper(Mapper[ImplementedResult, None, [CodeGenState]]):
```
```suggestion
class CodeGenMapper(Mapper[ImplementedResult, Never, [CodeGenState]]):
```

majosm replied:
Added here and in a few other places.

majosm commented:
Ready for another look I think @inducer.

inducer commented on pytato/target/python/numpy_like.py:
```diff
@@ -171,7 +172,7 @@ def _is_slice_trivial(slice_: NormalizedSlice,
 
 
-class NumpyCodegenMapper(CachedMapper[str, []]):
+class NumpyCodegenMapper(CachedMapper[str, FunctionDefinition, []]):
```
Is this annotation right? Does the numpy target generate appropriate code for functions?

majosm replied:
`NumpyCodegenMapper` doesn't support functions, so I guess this should be `Never`. Same for `FancyDotWriter`. Fixed these.

inducer commented on pytato/transform/calls.py:
```diff
@@ -58,6 +58,7 @@ class PlaceholderSubstitutor(CopyMapper):
     """
 
     def __init__(self, substitutions: Mapping[str, Array]) -> None:
+        # Ignoring function cache, not needed
```
How come? Not using the function cache may lead to large amounts of redundant work?

(I'm OK with not acting on this right now, but maybe this comment should be a FIXME?)

inducer commented on pytato/transform/__init__.py:
```diff
@@ -142,13 +139,15 @@
     A type variable representing the input type of a :class:`Mapper`.
 
-.. class:: CombineT
+.. class:: ResultT
 
-    A type variable representing the type of a :class:`CombineMapper`.
```
Did the `CombineT` docs get deleted on purpose?

majosm replied:
`CombineT` doesn't appear to exist anymore (looks like it got removed in #558).

inducer replied:
Oh! This is deleting stale docs. Got it, sorry about the noise.

inducer reviewed (commented):
Thanks! A few questions/comments below.

majosm commented on pytato/analysis/__init__.py:
```diff
@@ -413,16 +414,31 @@ class NodeCountMapper(CachedWalkMapper[[]]):
+    def __init__(
+            self,
+            count_duplicates: bool = False,
+            _visited_functions: set[Any] | None = None,
```
Looks like I did that to match `_visited_arrays_or_names` in `CachedWalkMapper`, which also uses `Any`. I guess I could change both to `Hashable`?

majosm commented:
I changed `CachedWalkMapper` to use an alias of `Hashable` instead of `Any`, and I removed a few other uses of `Any` that I found. Should be ready for another look.

inducer reviewed (approved):
Thanks! This looks great. In it goes.
